File: file-io/chunks/01-copy.go
package main

import (
	"fmt"
	"os"
	"strings"
	"syscall"
)

var BUF_SIZE int = 1024
var EXIT_FAILURE int = 1

func main() {
	var inputFd, outputFd int
	// var openFlags int
	// var filePerms uint32
	var numRead int
	// var buf [BUF_SIZE]byte - не сработало!
	buf := make([]byte, 2048)

	if len(os.Args) != 3 || strings.Compare(os.Args[1], "--help") == 0 {
		fmt.Println("too few arguments")
		fmt.Printf("%s old-file new-file\n", os.Args[0])
		os.Exit(EXIT_FAILURE)
	}

	inputFd, err := syscall.Open(os.Args[1], syscall.O_RDONLY, 0)
	if err != nil {
		fmt.Println("opening file:", err.Error())
		os.Exit(EXIT_FAILURE) // syscall.Exit(EXIT_FAILURE)
	}
	defer syscall.Close(inputFd)

	openFlags := syscall.O_CREAT | syscall.O_WRONLY | syscall.O_TRUNC
	filePerms := syscall.S_IRUSR | syscall.S_IWUSR | syscall.S_IRGRP |
		syscall.S_IWGRP | syscall.S_IROTH | syscall.S_IWOTH // rw-rw-rw

	// VS Code постоянно сам приводит filePerms к uint32
	outputFd, err = syscall.Open(os.Args[2], openFlags, uint32(filePerms)) // VS сам привел к uint32
	if err != nil {
		fmt.Println("opening file:", err.Error())
		os.Exit(EXIT_FAILURE)
	} else {
		fmt.Printf("file opened, fd=%d\n", outputFd)
	}
	defer syscall.Close(outputFd)
	
// --------------READ---------------------------------------------------
	numRead, err = syscall.Read(inputFd, buf)
	if err != nil {
		fmt.Println("reading file", err.Error())
	} else {
		fmt.Printf("read %d elements\n", numRead)
	}
// --------------WRITE---------------------------------------------------
	numWrite, err := syscall.Write(outputFd, buf)
	if err != nil {
		fmt.Println("wirting to file", err.Error())
	} else if numWrite != numRead {
		fmt.Println("couldn't write whole buffer")
	}
}
		// For check opened fd:
		// fmt.Println("opened fd: ", inputFd)
		// pid := strconv.Itoa(syscall.Getpid())
		// fmt.Printf("check open fd for %s PID:\n", pid)
		// path := "/proc/" + pid + "/fd"
		// cmd := exec.Command("ls", "-la", path)
		// out, err := cmd.Output()
		// if err != nil {
		// 	fmt.Println("output '$ls -la /proc/<PID>/fd' command")
		// 	syscall.Exit(EXIT_FAILURE)
		// } else {
		// 	fmt.Println(string(out))
		// }

File: file-io/chunks/01-copy.c
/* - open file only for read
*  - file for opening specified in command line arguments
*  - read data from file
*  - create new file if it is not exits
*  - write data to the new file
*/
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#ifndef BUF_SIZE        // Позволяет "cc -D" перекрыть определение
#define BUF_SIZE 1024
#endif

int
main(int argc, char *argv[]) {
    int inputFd, outputFd, openFlags;
    mode_t filePerms;
    ssize_t numRead;
    char buf[BUF_SIZE];

    // При запуске программы нужно ввести аргументы, проверяем их
    if(argc != 3 || strcmp(argv[1], "--help") == 0) {// если аргументов не 3 или второй - не хелп...
        printf("%s old-file new-file\n", argv[0]); 
    }

    // Открытие файлов ввода и вывода
    inputFd = open(argv[1], O_RDONLY);
    if(inputFd == -1) {
        perror("opening file");
        exit(EXIT_FAILURE);
    }

    openFlags = O_CREAT | O_WRONLY | O_TRUNC;
    filePerms = S_IRUSR | S_IWUSR  | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH; // rw-rw-rw-

    outputFd = open(argv[2], openFlags, filePerms);
    if(outputFd == -1) {
        perror("opening file");
        exit(EXIT_FAILURE);
    }

    // Перемещение данных до достиж-я конца файла ввода или возник-я ошибки:
    while( (numRead = read(inputFd, buf, BUF_SIZE)) > 0) {
        if(write(outputFd, buf, numRead) != numRead) {
            printf("could't write whole buffer");
        }
    }
    // при наступлении каких условий происходит выход из while? 
    
    if( numRead == -1 ) {
        perror("read");
        exit(EXIT_FAILURE);
    }
    if( close(inputFd) == -1 ) {
        perror("close input");
        exit(EXIT_FAILURE);
    }
    if( close(outputFd) == -1 ) {
        perror("close output");
        exit(EXIT_FAILURE);
    }

    exit(EXIT_SUCCESS);
}

// ./01-copy.exe /proc/self/maps maps.txt - копирование обычного файла
// ./01-copy.exe maps.txt /dev/tty - копировать обычный файл в это терминал

// read() returns zero if:
// - the file offset is at or past the end of file
// - no bytes are read
